shotover-proxy/src/transforms/mod.rs

```diff
@@ -475,6 +509,17 @@ pub trait Transform: Send {
     ///
     async fn transform<'a>(&'a mut self, message_wrapper: Wrapper<'a>) -> ChainResponse;
 
+    /// Called when shotover or buffered chain hosting the transform is shutting down.
+    /// The default implementation will automatically `call_next_shutdown` when non terminating.
+    /// So if you need to override this method to provide your own shutdown logic make sure to do the same.
+    async fn shutdown<'a>(&'a mut self, message_wrapper: Wrapper<'a>) -> ChainResponse {
```

Comment:
probably worth documenting what to expect about messages that are included in the method call here (like are these messages that came in before / after the shutdown hook?), do you absolutely need to handle them? If so it could cause a double up in logic being implemented in shutdown or encouraging people to call transform from shutdown to handle the messages provided in the shutdown hook.

I would actually be ok with us dropping any messages that were in the buffer as part of shutdown. Pretty much all clients will end up getting a stream closed error or a timeout and have their own trety logic. So we could get rid of the message_wrapper param here

Reply:
We also don't need to do it the same way transform does. (where wrapper has the list of transforms to call next). At this point I don't envisage one transform needing to know about the results of the downchain transforms shutdown result or need to know about it in an ergonomic way (the reason we do the whole Wrapper thing for the transform method).

We could just call shutdown on the transform chain and just have it go through the chains transforms one by one (or map or whatever) and call shutdown.

Reply:
I hit a few issues and completely rewrote this part, I'll update the PR description with the full details.